Refactor server game storage

khalidsharmarke opened the pull request:
Refactored game storage into class instances that allow for more modular and direct control over game state
edited socket validation to more accurately reflect which requests to drop

khalidsharmarke committed:
implemented server active game tracking

created new files - classes that outline how server interacts with client rooms and some supporting methods

khalidsharmarke committed:
Implemented turn-based drawing

server now locks out indiviual clients from interacting with their sketchboards while allowing  the player whose turn it is, to submit new images

willwen commented on client/scripts/index.js:
```diff
@@ -20,6 +22,20 @@ function getPlayerColorFromNumber(player_number) {
 
 }
 
+function enableButtons(data){
```
why does this take the argument data ?

willwen replied:
this also seems to modify the whiteboard
soo
maybe call the function
enableButtonsAndBoard

willwen commented on client/scripts/index.js:
```diff
@@ -1,3 +1,5 @@
+let pad = null
```
you can put a TODO here 
you made it global to ignore scoping
but eventually you can pass around the pad to each function that needs it to control the scope

willwen replied:
and add a trello ticket to refactor `pad`'s scoping

willwen commented on client/scripts/index.js:
```diff
@@ -69,18 +86,19 @@ function game_init(pad) {
         player_number: player_number,
         new_board_state: encodeURIComponent(newBoardState)
     })
+    disableButtons()
```
can add a comment here, to explain in plain english
that after you hit submit, we disable all buttons for the user  until its their turn again

willwen commented on client/scripts/index.js:
```diff
@@ -69,18 +86,19 @@ function game_init(pad) {
 function update_image_from_server(data) {
     console.log('new image received')
-    console.log(data)
+    let thing = decodeURIComponent(data.new_board_state)
```
horrible variable name

willwen commented on client/scripts/index.js:
```diff
@@ -92,6 +110,11 @@ requirejs(['responsive-sketchpad/sketchpad'],
 )
 
+function displayRoomID(data){
+    console.log(data)
```
no need to log this anymore

willwen replied:
why is it called data
the param should be called room_id

willwen commented on client/scripts/index.js:
```diff
@@ -108,4 +131,7 @@ socket.on('disconnect', reason => {
     window.location = '/'
 })
 
```
can add a comment here explaining this is where you do all the socket handlers:
ex:
```
// action handlers after receiving messages from server:
socket.on('room_id', displayRoomID)
socket.on('new_image', update_image_from_server)
socket.on('player_number', disableButtons)
socket.on('enable_turn', enableButtons)
```

willwen commented on index.js:
```diff
@@ -1,8 +1,10 @@
 const PORT = process.env.PORT || 8000
 
+const {set_of_active_games, generate_game_id, checkIfGameAlive, createNewGame, getGameFromSet} = require('./utils/game')
```
inconsistent variable naming
is it `_` or Snake case ?

willwen commented on index.js:
```diff
@@ -44,43 +45,20 @@ app.use('/', function (req, res, next) {
-		// game exists
+
+    const gameAlive = checkIfGameAlive(game_id_to_join)
+    if (gameAlive){
```
the spacing and tab formatting  is really messed up here

willwen commented on index.js:
```diff
@@ -108,30 +86,38 @@ function getRoomFromSocket(socket_obj){
 function validateSocket(socket_obj){
     const room_id = getRoomFromSocket(socket_obj)
-
+    const room = getGameFromSet(room_id)
```
technically dont want const here
you alter call room.addPlayer
which manipulates the room's player field

willwen replied:
`getGameFromSet` is unclear
Set isnt clear enough
I would do:
`getGameObjFromRoomId`
or `getGameObj()`

willwen commented on index.js:
```diff
@@ -108,30 +86,38 @@ function getRoomFromSocket(socket_obj){
+    if (room_id !== null && !socket_obj.rooms.has(room_id) && room !== null){
         socket_obj.join(room_id)
+        room.addPlayer(socket_obj.id)
+        return {room_id, room}
```
single return point
so result = {room_id, room}


result = null


return result

willwen commented on index.js:
```diff
@@ -108,30 +86,38 @@ function getRoomFromSocket(socket_obj){
-    // TODO:
-    // here we should instantiate a game obj
-    // to track game state and stack history
+    const {room_id, room} = validateSocket(socket)
```
this function can return null, but you dont handle the null case ?

willwen commented on index.js:
```diff
@@ -108,30 +86,38 @@ function getRoomFromSocket(socket_obj){
+    io.to(room.getCurrentPlayer()).emit('enable_turn', true)
 
     // update users in room for new image
     socket.on('gameplay_stroke', data => {
```
in this case data is the image data right ? 
so maybe jsut call it
`json_image_data`

willwen commented on utils/ActiveGames.js:
```diff
@@ -0,0 +1,10 @@
+class ActiveGames{
+    add(room_id, room_obj){
```
it should initialize an empty dictionary

```
class ActiveGames {
     activeGamesSet = {}


    add ( ) {
    ...
    }
}
```

willwen replied:
like in here
https://github.com/khalidsharmarke/A-Fake-Artist-Goes-To-New-York/pull/9/files#diff-6ce89e5b021636a347e8d3ed4e43055c9ba2357870bfa913c76c796dd6f68c40R4

khalidsharmarke replied:
i disagree. this becomes redundant.
given the scope of the class as an implementation to store rooms in a singular obj
i think it best to avoid having to write ActiveGames.rooms['insert room id here']

willwen replied:
look up `extends object`

willwen commented on utils/Room.js:
```diff
@@ -0,0 +1,31 @@
+module.exports = class Room{
+    #private = 123
```
?why is this here

khalidsharmarke replied:
i was experimenting with private class fields
its been removed

willwen commented on utils/Room.js:
```diff
@@ -0,0 +1,31 @@
+        this.listOfPlayers = []
+        this.nextPlayer = 0;
+    }
+    incrementPlayer(){
```
its incrementing the player number not the player
make the function name better

willwen commented on utils/Room.js:
```diff
@@ -0,0 +1,31 @@
+    #private = 123
+    constructor(){
+        this.listOfPlayers = []
+        this.nextPlayer = 0;
```
nextPlayerNumber

willwen commented on utils/Room.js:
```diff
@@ -0,0 +1,31 @@
+            this.nextPlayer = 0
+        }
+    }
+    nextTurn(){
```
next turn does not sound like it returns anything, so it should return null

you should make a separate function that gets Player

so you do:
```
game.nextTurn();
nextPlayerNumber = game.getNextPlayerNumber()
```

notice how the function names describes exactly what is going on

willwen commented on utils/Room.js:
```diff
@@ -0,0 +1,31 @@
+        }
+    }
+    nextTurn(){
+        const player = this.listOfPlayers[this.nextPlayer]
```
this is the player's socket, not the player

willwen commented on utils/Room.js:
```diff
@@ -0,0 +1,31 @@
+    getCurrentPlayer(){
+        return this.listOfPlayers[this.nextPlayer]
+    }
+    test(){
```
what

willwen commented on utils/Room.js:
```diff
@@ -0,0 +1,31 @@
+        return player
+    }
+    addPlayer(player_as_socket){
+        this.listOfPlayers.push(player_as_socket)
```
why player_as_socket

instead of 
`playerSocketId`

willwen commented on utils/game.js:
```diff
@@ -0,0 +1,42 @@
+const set_of_active_games = new ActiveGames()
+
+
+function generate_game_id(){
```
fix camel or snake case

willwen reviewed (commented):
lots of naming issues

khalidsharmarke committed:
Refactored Function Names

changed function names for more clarity

khalidsharmarke committed:
Refactored Room Method Arguments

khalidsharmarke committed:
added line at end of file

willwen commented on client/scripts/index.js:
```diff
@@ -53,59 +72,61 @@ function sketchpad_init(Sketchpad) {
-    console.log('new image received')
-    console.log(data)
+// overwrites current pad with JSONimage from server
+function updateImageFromServer(data) {
```
data -> whiteboard_json_data

willwen commented on index.js:
```diff
@@ -28,59 +29,31 @@ const cookieOptions = {
-	console.log(list_of_active_games)
-
-    // Set cookie
+    // Set cookie`
```
typo

willwen replied:
why is this resolved if the typo is still there

willwen commented on index.js:
```diff
@@ -89,11 +62,11 @@ app.post('/join_game', urlencodedParser, (req, res)=>{
+function getRoomIDFromSocket(socket_obj){
     try{
         const parsed_cookie = cookie.parse(socket_obj.request.headers.cookie)
         const room_id = Number(parsed_cookie['room_id'])
```
single return statements

willwen commented on index.js:
```diff
@@ -107,31 +80,43 @@ function getRoomFromSocket(socket_obj){
-    const room_id = getRoomFromSocket(socket_obj)
-
+    let result = null
+    const room_id = getRoomIDFromSocket(socket_obj)
```
?

willwen commented on index.js:
```diff
@@ -107,31 +80,43 @@ function getRoomFromSocket(socket_obj){
-    // here we should instantiate a game obj
-    // to track game state and stack history
+    // validate all incoming connections
+    const {room_id, room} = validateSocket(socket)
```
getRoomFromSocket()

willwen replied:
room id should be a field on room obj

willwen replied:
move redundant 
`room.getCurrentPlayer()`
up at top

willwen commented on index.js:
```diff
@@ -107,31 +80,43 @@ function getRoomFromSocket(socket_obj){
+    // validate all incoming connections
+    const {room_id, room} = validateSocket(socket)
+    // stops server-side operations on invalid socket connection
+    if (room_id == null || room == null) return
```
only need room==null check

willwen commented on index.js:
```diff
@@ -107,31 +80,43 @@ function getRoomFromSocket(socket_obj){
     });
 });
-   
+   
```
check your newlines

willwen commented on utils/Room.js:
```diff
@@ -0,0 +1,31 @@
+module.exports = class Room{
```
export at the end of the file

willwen commented on utils/Room.js:
```diff
@@ -0,0 +1,31 @@
+module.exports = class Room{
+    constructor(){
+        this.listOfPlayers = []
+        this.nextPlayerNumber = 0
```
get current player number in

and next player number can be derived from current player

willwen commented on utils/game.js:
```diff
@@ -0,0 +1,39 @@
+}
+
+function checkIfGameExists(game_id){
+    return setOfActiveGames.has(game_id)
```
game_id vs room_id? make it consistent

willwen commented on utils/game.js:
```diff
@@ -0,0 +1,39 @@
+}
+
+function getRoomFromActiveSet(game_id){
+    if (setOfActiveGames.has(game_id)){
```
checkIfGameExists

willwen commented on utils/game.js:
```diff
@@ -0,0 +1,39 @@
+    try {
+        const game_id = generateGameId()
+        let game = new Room(game_id)
+        setOfActiveGames.add(game_id, game)
```
jsut add a game obj

khalidsharmarke committed:
Implemented better function args

willwen commented on client/scripts/index.js:
```diff
@@ -19,9 +23,23 @@ function getPlayerColorFromNumber(player_number) {
     return '#' + colors[player_number]
 
 }
+// allows player interations with interface
```
your PR still doesnt have a clear title
and doesnt have a body

willwen commented on client/scripts/index.js:
```diff
@@ -19,9 +23,23 @@ function getPlayerColorFromNumber(player_number) {
 
 }
+// allows player interations with interface
+function enablePlayerActions(data){
```
why is data in the args

khalidsharmarke replied:
i think it important to denote that the function is able to receive data from the server.
Im still thinking theres an implementation where the server passes the pad object back to the client
with that being how we control scope

willwen commented on client/scripts/index.js:
```diff
@@ -53,59 +72,61 @@ function sketchpad_init(Sketchpad) {
 // after a player has submitted their stroke,
 // this function exports the whiteboard as JSON and sends it to the server
-function submit_stroke_to_server(player_number, pad) {
+function submitStrokeToServer(player_number, pad) {
```
this function makes it sound like you are passing a stroke to the server
but the arg is pad
so either change the function to be
submit pad to server
OR
modify the param to be the strokeJson data instead of pad

willwen commented on index.js:
```diff
@@ -107,31 +80,43 @@ function getRoomFromSocket(socket_obj){
-
+    let result = null
+    const room_id = getRoomIDFromSocket(socket_obj)
+    const room = getRoomFromActiveSet(room_id)
```
`getRoomObjFromID`

willwen commented on index.js:
```diff
@@ -107,31 +80,43 @@ function getRoomFromSocket(socket_obj){
 function validateSocket(socket_obj){
-    const room_id = getRoomFromSocket(socket_obj)
-
+    let result = null
```
this function doesnt validate socket
it returns a room

willwen commented on index.js:
```diff
@@ -107,31 +80,43 @@ function getRoomFromSocket(socket_obj){
+    if (room_id !== null && !socket_obj.rooms.has(room_id) && room !== null){
         socket_obj.join(room_id)
+        room.addPlayer(socket_obj.id)
+        result = {room_id, room}
```
why return both room_id and room

willwen commented on utils/Room.js:
```diff
@@ -0,0 +1,31 @@
+        return this.listOfPlayers.indexOf(player_as_socket_id)
+    }
+    getCurrentPlayer(){
+        return this.listOfPlayers[this.nextPlayerNumber]
```
i dont like the variable names here
isee you are leveraging the fact that arrays are 0 indexed,
but it should really look like:
```
getCurrentPlayer(){
     return this.listOfCurerntPlayers[this.currentPlayerNumber + 1 ] // +1 because array index starts at 0
}
```

khalidsharmarke committed:
Refactored function names to standard

khalidsharmarke committed:
Merge branch 'refactor-server-game-storage' of https://github.com/khalidsharmarke/A-Fake-Artist-Goes-To-New-York into refactor-server-game-storage

# Conflicts:
#	utils/Room.js

khalidsharmarke committed:
Hotfix involving player turns not working

had to perform  getCurrentPlayerSocket() twice due to room state change during image submission from client
